shell/common/api/electron_api_asar.cc

```diff
@@ -103,15 +107,67 @@ class Archive : public gin_helper::Wrappable<Archive> {
     return gin::ConvertToV8(isolate, new_path);
   }
 
-  // Return the file descriptor.
-  int GetFD() const {
-    if (!archive_)
-      return -1;
-    return archive_->GetFD();
+  v8::Local<v8::ArrayBuffer> ReadSync(gin_helper::ErrorThrower thrower,
+                                      uint64_t offset,
+                                      uint64_t length) {
+    base::CheckedNumeric<uint64_t> safe_offset(offset);
+    base::CheckedNumeric<uint64_t> safe_end = safe_offset + length;
+    if (!safe_end.IsValid() ||
+        safe_end.ValueOrDie() > archive_->file()->length()) {
+      thrower.ThrowError("Out of bounds read requested in ASAR");
+      return v8::Local<v8::ArrayBuffer>();
+    }
+    auto array_buffer = v8::ArrayBuffer::New(thrower.isolate(), length);
+    auto backing_store = array_buffer->GetBackingStore();
+    memcpy(backing_store->Data(), archive_->file()->data() + offset, length);
+    return array_buffer;
+  }
+
+  v8::Local<v8::Promise> Read(gin_helper::ErrorThrower thrower,
+                              uint64_t offset,
+                              uint64_t length) {
+    gin_helper::Promise<v8::Local<v8::ArrayBuffer>> promise(thrower.isolate());
+    v8::Local<v8::Promise> handle = promise.GetHandle();
+
+    base::CheckedNumeric<uint64_t> safe_offset(offset);
+    base::CheckedNumeric<uint64_t> safe_end = safe_offset + length;
+    if (!safe_end.IsValid() ||
+        safe_end.ValueOrDie() > archive_->file()->length()) {
+      thrower.ThrowError("Out of bounds read requested in ASAR");
+      return v8::Local<v8::Promise>();
+    }
+
+    base::ThreadPool::PostTaskAndReplyWithResult(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::USER_VISIBLE},
+        base::BindOnce(&Archive::ReadOnIO, thrower.isolate(), archive_, offset,
+                       length),
+        base::BindOnce(&Archive::ResolveReadOnUI, std::move(promise)));
+
+    return handle;
   }
 
  private:
-  std::unique_ptr<asar::Archive> archive_;
+  static std::unique_ptr<v8::BackingStore> ReadOnIO(
+      v8::Isolate* isolate,
+      std::shared_ptr<asar::Archive> archive,
+      uint64_t offset,
+      uint64_t length) {
+    auto backing_store = v8::ArrayBuffer::NewBackingStore(isolate, length);
```

Comment:
This API is not marked as thread safe, and [its implementation](https://source.chromium.org/chromium/chromium/src/+/master:v8/src/objects/backing-store.cc;drc=65c53735b58c961c225a878df11c841a6614d6ab;l=219?originalUrl=https:%2F%2Fcs.chromium.org%2F) does access things from `isolate`. I think the `backing_store` should be created in UI thread and then passed to IO.

Reply:
The ArrayBuffer allocator is noted to be thread-safe: https://source.chromium.org/chromium/chromium/src/+/master:v8/include/v8.h;l=5041;drc=76b53b66b0a441494748b82250b7d3ea3f1ef6fe. Additionally, SharedArrayBuffer::NewBackingStore is surely thread-safe, as that is the entire reason for the existence of SharedArrayBuffer, and it [calls the same allocation function](https://source.chromium.org/chromium/chromium/src/+/master:v8/src/api/api.cc;l=7890;drc=76b53b66b0a441494748b82250b7d3ea3f1ef6fe).

Reply:
Yeah the allocator is thread-safe and it is safe to share the backing stores between threads, but I'm not sure if the `SharedArrayBuffer::NewBackingStore` API is thread-safe.

From my understanding of V8's threading model, the isolate should not be used in multiple threads without locks. In Node.js, the shared array buffers (and their backing stores) are always created in the same thread with their isolates, and only then their backing stores can be passed to other threads. It is similar to how we can write to a `node::Buffer` in a new thread, while still having to create it in the thread of the isolate.

Also in the [the implementation of BackingStore::Allocate](https://source.chromium.org/chromium/chromium/src/+/master:v8/src/objects/backing-store.cc;drc=65c53735b58c961c225a878df11c841a6614d6ab;l=219?originalUrl=https:%2F%2Fcs.chromium.org%2F), it has a call of `isolate->heap()->AllocateExternalBackingStore`, which could invoke V8's garbage collection. It should not be a thread-safe API.

Reply:
makes sense, thanks for the info. updated.